File: src/main/java/com/refinedmods/refinedstorage/inventory/item/validator/UpgradeItemValidator.java
package com.refinedmods.refinedstorage.inventory.item.validator;

import com.refinedmods.refinedstorage.item.UpgradeItem;
import net.minecraft.world.item.ItemStack;

import java.util.function.Predicate;

public class UpgradeItemValidator implements Predicate<ItemStack> {
    private final UpgradeItem.Type type;

    public UpgradeItemValidator(UpgradeItem.Type type) {
        this.type = type;
    }

    @Override
    public boolean test(ItemStack stack) {
        return stack.getItem() instanceof UpgradeItem && ((UpgradeItem) stack.getItem()).getType() == type;
    }
}

File: src/main/kotlin/com/refinedmods/refinedstorage/inventory/item/validator/UpgradeItemValidator.kt
package com.refinedmods.refinedstorage.inventory.item.validator

import com.refinedmods.refinedstorage.item.UpgradeItem
import net.minecraft.item.ItemStack
import java.util.function.Predicate

class UpgradeItemValidator(private val type: UpgradeItem.Type) : Predicate<ItemStack> {
    override fun test(stack: ItemStack): Boolean {
        return stack.item is UpgradeItem && (stack.item as UpgradeItem).type == type
    }
}